pkg/runtime/runtime.go

```diff
@@ -1749,6 +1748,9 @@ func (a *DaprRuntime) initPubSub(c componentsV1alpha1.Component) error {
 // And then forward them to the Pub/Sub component.
 // This method is used by the HTTP and gRPC APIs.
 func (a *DaprRuntime) Publish(req *pubsub.PublishRequest) error {
+	a.topicsLock.RLock()
```

Comment:
This is likely to have adverse performance implications. If races here are rare (for example, occuring only when someone is trying to publish a message while processing subscriptions on init), I'd remove this.

Reply:
It can happen every time the subscriptions are changed, for example if apphealth is turning pubsub off.

Although this will negatively impact performance a bit due to the overhead of acquiring a lock, I am using a RWMutex with a read-lock here, so it doesn't block if other goroutines are trying to publish at the same time. It will only block if there's an operation that's modifying the topics/subscriptions concurrently.

What do you think?

Reply:
Is there a way to have this map to be immutable?

Reply:
I prefer the correctness over the performance here. BTW, we don't have performance tests for pubsub (yet).

Reply:
At the very least then lock when accessing the map and not the entire function?

Reply:
That sounds very reasonable. Will do

Reply:
> I prefer the correctness over the performance here. BTW, we don't have performance tests for pubsub (yet).

We actually do have pub/sub perf tests, courtesy of @skyao